### PR TITLE
High velocity filter

### PR DESCRIPTION
**Pull request by alhighfidelity**
Build to run real time test for filter pass through.

### CODE REVIEW

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -347,7 +347,7 @@ class DeadlockWatchdogThread : public QThread {
                 // Don't actually crash in debug builds, in case this apparent deadlock is simply from
                 // the developer actively debugging code
                 #ifdef NDEBUG
-                    deadlockDetectionCrash();
+                    //deadlockDetectionCrash();
```
Comment (hyperlogic) [inline interface/src/Application.cpp]: Please uncomment deadlockDetectionCrash()

File: libraries/controllers/src/controllers/InputRecorder.cpp [hunk 2]
```diff
@@ -313,7 +313,7 @@ namespace controller {
         }
 
         if (_playback) {
-            _playCount++;
+           _playCount++;
```
Comment (hyperlogic) [inline libraries/controllers/src/controllers/InputRecorder.cpp]: ident

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 3]
```diff
@@ -396,6 +396,65 @@ void ViveControllerManager::InputDevice::emitCalibrationStatus(const bool succes
...
+}
+
+
+void ViveControllerManager::InputDevice::processHighVelocityFilter(){
+
+
+
+
+
+}
```
Comment (hyperlogic) [inline plugins/openvr/src/ViveControllerManager.cpp]: All this code was for testing right? you should remove it, now that it has been moved into the HighVelocity filter.

File: interface/resources/controllers/standard.json [hunk 4]
```diff
@@ -61,7 +61,11 @@
         { "from": "Standard.RightHand", "to": "Actions.RightHand" },
 
         { "from": "Standard.LeftFoot", "to": "Actions.LeftFoot" },
-        { "from": "Standard.RightFoot", "to": "Actions.RightFoot" },
+
+         {
+          "from": "Standard.RightFoot", "to" : "Actions.RightFoot",
+            "filters" : [{"type" :  "highVelocity", "p_threshold" : 0.5, "p_weight": 2, "size": 5, "q_threshold" : 0.5, "q_weight" : 2}]
+        },
```
Comment (hyperlogic) [inline interface/resources/controllers/standard.json]: we are only mapping the right foot? what about the other foot?  what about hips, chest etc.

File: libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+           // pThreshold(d_pos,pos);
+
+           
+            // rotation processing
+
+            //glm::quat dQ = deltaQ(rot);
+            //float q_dot = qDot(_rotOutput[_pCount - 1], dQ);
+            //qThreshold(q_dot, _qThresh);
+
+        //}
```
Comment (hyperlogic) [inline libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp]: All of these debug comments should be remove or put behind a #ifdef WANT_DEBUG preprocessor guard

File: libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+
+        glm::quat qOut = updateRotOut(rot);
+
+        
+       Pose ret = DataToPose(pos, rot);
+
+        // print output values
+
+       qDebug() << "Output: " << pos.x << " " << pos.y << " " << pos.z
+           << " " << rot.w << " " << rot.x << " " << rot.y << " " << rot.z << endl;
```
Comment (hyperlogic) [inline libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp]: Please remove, or put behind a #ifdef WANT_DEBUG guard.

File: libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+            _rotOutput.push_back(qtmp);
+        }
+
+    }
+
+    glm::vec3 HighVelocityFilter::updatePosOut(glm::vec3 v) const {
+
+        glm::vec3 ret;
+        
+        qDebug() << "Input: N = " << _posOutput.size() << " " << v.x << " " << v.y << " " << v.z;
```
Comment (hyperlogic) [inline libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp]: Please remove, or put behind a #ifdef WANT_DEBUG guard.

File: libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+        }
+
+        // N should be equal to _n
+
+        ret = _posOutput[0];
+        std::vector<glm::vec3>::iterator it = _posOutput.begin();
+        _posOutput.erase(it);
+        _posOutput.push_back(v);
+
+        qDebug() << "Output: N = " << _posOutput.size() << " " << ret.x << " " << ret.y << " " << ret.z;
```
Comment (hyperlogic) [inline libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp]: Please remove, or put behind a #ifdef WANT_DEBUG guard.

**Comment by Atlante45** [comment]
test this please

**Comment by alhighfidelity** [comment]
This is for testing only. Thanks for the feedback though. I can remove the
filter from the ViveManager class.

File: interface/resources/controllers/standard.json (hunk 4 above)
Comment (alhighfidelity) [inline interface/resources/controllers/standard.json]: This is to test the right foot only to test algorithm. Other limbs can easily be added after the testing.

File: libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp (hunk 5 above)
Comment (alhighfidelity) [inline libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp]: This is to debug the algorithm - these are actually parts of the filter that have been commented out to test how the input is behaving in real time. The avatar is not moving properly with input data coming through the InputRecorder class.

File: libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp (hunk 6 above)
Comment (alhighfidelity) [inline libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.cpp]: These are for debugging as well.

**Comment by AndrewMeadows** [comment]
The ubuntu build failed for the following error:
```
00:54:40 /home/jenkins/workspace/pr-build/label/ubuntu/libraries/controllers/src/controllers/impl/filters/HighVelocityFilter.h:46:22: error: extra qualification ‘controller::HighVelocityFilter::’ on member ‘qThreshold’ [-fpermissive]
00:54:40                  void HighVelocityFilter::qThreshold(const float &signal, const float &threshold) const;
```
Remove the `HighVelocityFilter::` namespace qualifier when already in the namespace.

**Comment by alhighfidelity** [comment]
Put in a new commit and pull request.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1604/HighFidelity-Beta-PR10946-cb81158073d88e492119f93dc67c5c7001be2db2.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1604/HighFidelity-Beta-PR10946-cb81158073d88e492119f93dc67c5c7001be2db2.dmg

**Pull request closed without merging**